### PR TITLE
[Table] Adds focus cell tab and enter microinteraction behavior

### PR DESCRIPTION
**Pull request by gscshoyru**
#### Checklist

- [x] Include tests

#### Changes proposed in this pull request:
Focus cells now track what region they belong to.

Pressing enter/tab and their shift-ed counterparts move the cell inside the selected region

#### Reviewers should focus on:
There's a... large block of code in this code that's duplicated a bunch, with minor changes. There really should be a way to simplify it. I'm not sure what that is. Help, please?

Also, all of my names for functions are pretty mediocre here. 

#### Screenshot

jun-19-2017 17-13-53

### CODE REVIEW

**Comment by cmslewis** [comment]
The interactions feel great! Only thing I noticed was that editable cells ruin the party once again:
tab-focus

File: packages/table/src/table.tsx [hunk 1]
```diff
@@ -1265,6 +1302,171 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
         this.scrollBodyToFocusedCell(newFocusedCell);
     }
 
+    // no good way to call arrow-key keyboard events from tests
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: There should be a way forward with this now! See `tableTests.tsx#L500`

File: packages/table/src/table.tsx [hunk 2]
```diff
@@ -1265,6 +1302,171 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+                return;
+            }
+
+            const focusCellRegion = Regions.getCellRegionFromRegion(
+                selectedRegions[focusedCell.focusSelectionIndex],
+                grid.numRows,
+                grid.numCols,
+            );
+
+            if (focusCellRegion.cols[0] === focusCellRegion.cols[1] &&
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: nit: elsewhere in the codebase, I think the pattern is becoming

```
if (A
    && B
    && C) {

}
```

File: packages/table/src/table.tsx [hunk 3]
```diff
@@ -1265,6 +1302,171 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+                focusCellRegion.rows[0] === focusCellRegion.rows[1] &&
+                selectedRegions.length === 1) {
+                this.handleFocusMove(e, direction);
+                return;
+            }
+
+            switch (direction) {
+                case "up":
+                    newFocusedCell.row -= 1;
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: For all this repeated code, I goofed around and got to this, which might be more abstracted than it's worth:

```
function moveFocusCell(primaryAxis: "row" | "col", secondaryAxis: "row" | "col", isUpOrLeft: boolean) {
    const primaryAxisPlural = primaryAxis === "row" ? "rows" : "cols";
    const secondaryAxisPlural = secondaryAxis === "row" ? "rows" : "cols";

    const generalDelta = isUpOrLeft ? -1 : +1;
    const generalIndex = isUpOrLeft ? 1 : 0;

    newFocusedCell[primaryAxis] += generalDelta;

    const isPrimaryIndexOutOfBounds = isUpOrLeft
        ? newFocusedCell[primaryAxis] < focusCellRegion[primaryAxisPlural][0]
        : newFocusedCell[primaryAxis] > focusCellRegion[primaryAxisPlural][1];

    if (isPrimaryIndexOutOfBounds) {
        newFocusedCell[primaryAxis] = focusCellRegion[primaryAxisPlural][generalIndex];
        newFocusedCell[secondaryAxis] += generalDelta;

        const isSecondaryIndexOutOfBounds = isUpOrLeft
            ? newFocusedCell[secondaryAxis] < focusCellRegion[secondaryAxisPlural][0]
            : newFocusedCell[secondaryAxis] > focusCellRegion[secondaryAxisPlural][1];

        if (isSecondaryIndexOutOfBounds) {
            let newFocusCellSelectionIndex = focusedCell.focusSelectionIndex + generalDelta;
            
            if (isUpOrLeft
                ? newFocusCellSelectionIndex < 0
                : newFocusCellSelectionIndex >= selectedRegions.length
            ) {
                newFocusCellSelectionIndex = isUpOrLeft
                    ? selectedRegions.length - 1
                    : 0;
            }
            
            const newFocusCellRegion = Regions.getCellRegionFromRegion(
                selectedRegions[newFocusCellSelectionIndex],
                grid.numRows,
                grid.numCols,
            );

            newFocusedCell = {
                col: newFocusCellRegion.cols[generalIndex],
                focusSelectionIndex: newFocusCellSelectionIndex,
                row: newFocusCellRegion.rows[generalIndex],
            };
        }
    }
}
```

You could then call this function as follows:

```
switch (direction) {
    case "up":
        moveFocusCell("row", "col", true);
    case "left":
        moveFocusCell("col", "row", true);
    case "down":
        moveFocusCell("row", "col", false);
    case "right":
        moveFocusCell("col", "row", false);
}
```

File: packages/table/src/table.tsx [hunk 4]
```diff
@@ -1265,6 +1302,171 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+            }
+        }
+
+        if (newFocusedCell.row < 0 || newFocusedCell.row >= grid.numRows ||
+            newFocusedCell.col < 0 || newFocusedCell.col >= grid.numCols) {
+            return;
+        }
+
+        // change selection to match new focus cell location
+        // const newSelectionRegions = [Regions.cell(newFocusedCell.row, newFocusedCell.col)];
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: commented code

File: packages/table/src/table.tsx (hunk 1 above)
Comment (gscshoyru) [inline packages/table/src/table.tsx]: Bad copy/paste, sorry, shouldn't have taken the comment with me

File: packages/table/src/table.tsx (hunk 3 above)
Comment (gscshoyru) [inline packages/table/src/table.tsx]: I mean, I _could_ . 

This code is very... abstract, as you said, though. And abstracting this much makes it super hard to follow. 

I'll see if I can't find some middle ground, or something. Or possibly just comment it to death.

File: packages/table/src/table.tsx [hunk 5]
```diff
@@ -586,6 +586,72 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
     }
 
+    private moveFocusCell(
+        primaryAxis: "row" | "col",
+        secondaryAxis: "row" | "col",
+        isUpOrLeft: boolean,
+        newFocusedCell: IFocusedCellCoordinates,
+        focusCellRegion: IRegion,
+    ) {
+
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: nit: don't really need this newline

File: packages/table/src/table.tsx [hunk 6]
```diff
@@ -586,6 +586,72 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+                newFocusedCell = {
+                    col: newFocusCellRegion.cols[regionIntervalIndex],
+                    focusSelectionIndex: newFocusCellSelectionIndex,
+                    row: newFocusCellRegion.rows[regionIntervalIndex],
+                };
+            }
+        }
+        return newFocusedCell;
+    }
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: Not sure we saved much in terms of lines of code, but at least it's not redundant!

File: packages/table/test/tableTests.tsx [hunk 7]
```diff
@@ -519,6 +519,65 @@ describe("<Table>", () => {
             }
         });
 
+        describe("moves a focus cell interally with tab and enter", () => {
```
Comment (cmslewis) [inline packages/table/test/tableTests.tsx]: spellcheck: *inter*-**n**-*ally*
